src/IK/include/BipedalLocomotion/IK/InverseKinematics.h

```diff
@@ -0,0 +1,96 @@
+/**
+ * @file InverseKinematics.h
+ * @authors Giulio Romualdi
+ * @copyright 2020 Istituto Italiano di Tecnologia (IIT). This software may be modified and
+ * distributed under the terms of the GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#ifndef BIPEDAL_LOCOMOTION_IK_INVERSE_KINEMATICS_H
+#define BIPEDAL_LOCOMOTION_IK_INVERSE_KINEMATICS_H
+
+#include <memory>
+#include <string>
+#include <optional>
+
+#include <Eigen/Dense>
+#include <manif/SE3.h>
+
+#include <BipedalLocomotion/ParametersHandler/IParametersHandler.h>
+#include <BipedalLocomotion/System/Advanceable.h>
+#include <BipedalLocomotion/IK/Task.h>
+
+namespace BipedalLocomotion
+{
+
+namespace IK
+{
+
+/**
+ * State of the InverseKinematics
+ */
+struct IKState
+{
+    Eigen::VectorXd jointVelocity; /**< Joints velocity in rad per seconds */
+    manif::SE3d::Tangent baseVelocity; /**< Mixed spatial velocity of the base */
+};
+
+/**
+ * InverseKinematics implements the interface for the inverse kinematics. Please inherits this class
+ * if you want to implement your custom IK.
+ */
+class InverseKinematics : public BipedalLocomotion::System::Advanceable<IKState>
+{
+
+public:
+    /**
+     * Initialize the planner.
+     * @param handler pointer to the parameter handler.
+     * @return true in case of success/false otherwise.
+     */
+    virtual bool initialize(std::weak_ptr<ParametersHandler::IParametersHandler> handler);
+
+    /**
+     * Add a task in the inverse kinematics
+     * @param task pointer to a given task
+     * @param priority is the priority of the task. The lower the number the higher the priority.
+     * @param weight weight associated to the task.
+     * @param isInequality true if the task should be treated as inequality constraint
+     * @return true if the task has been added to the inverse kinematics.
+     */
+    virtual bool addTask(std::shared_ptr<Task> task,
+                         const std::string& taskName,
+                         std::size_t priority,
+                         std::optional<Eigen::Ref<const Eigen::VectorXd>> weight = {},
+                         bool isInequality = false) = 0;
```

Comment:
Ideally, the information on whether the task is an equality or not belongs to the task itself. I would consider moving this to the task class itself. In addition, for readability purposes, I would avoid using booleans when their meaning is not evident https://abseil.io/tips/94

Reply:
Also, how are inequalities handled if they are not constraints?

Reply:
> Ideally, the information on whether the task is an equality or not belongs to the task itself. I would consider moving this to the task class itself.

I agree 


> I would avoid using booleans when their meaning is not evident https://abseil.io/tips/94

I can create a struct that contains the priority and the weight and pass it to the ik every time a new task is handled 

> Also, how are inequalities handled if they are not constraints?

In the specific QP implementation, this is not supported. I can return an error if the user wants to do that

Reply:
>Also, how are inequalities handled if they are not constraints?

More relevant thread: Also, how are inequalities handled if they are not constraints?

Reply:
> I can create a struct that contains the priority and the weight and pass it to the ik every time a new task is handled

If you move it inside the ``Task`` class, I think that is enough. For what concerns the priority, I like how it is done right now. It makes a lot of sense to set the priority only when adding the task, as it is more related to own the task is considered together with the others rather than the task itself.

Reply:
Done in 56b575c33a0152d1774a91b545415252cfd036a3 421043c08b12d00e53bcbfa73e1a1d40158cc5c0 da394d9231983063376294518ff6cdb602867bb1 e049f4e899bea387162621ea8380da52f84b09f3 2453f9e199d2f8866d637864ef63a5c1293c293d f0cee977e288508bc666a9180b4816f78db2356d